EIPS/pairings.md

```diff
@@ -0,0 +1,138 @@
+
+## Specification
+
+Add a precompiled contracts for a bilinear function on groups on the elliptic curve "alt_bn128". We will define the precompiled contract in terms of a discrete logarithm. The discrete logarithm is of course assumed to be hard to compute, but we will give an equivalent specification that makes use of elliptic curve pairing functions which can be efficiently computed below.
+
+Address: 0x8
+
+For a cyclic group `G` (written additively) of prime order `q` let `log_P: G -> F_q` be the discrete logarithm on this group with respect to a generator `P`, i.e. `log_P(x)` is the smallest non-negative integer `n` such that `n * P = x`.
+
+The precompiled contract is defined as follows, where the two groups `G_1` and `G_2` and their generators `P_1` and `P_2` are defined below. Both generators have the same prime order `q` and the actual choice of the generators does not matter, as long as they have order `q`.
+
+```
+Input: (a1, b1, a2, b2, ..., ak, bk) from (G_1 x G_2)^k
+Output: If the length of the input is incorrect or any of the inputs are not elements of
+        the respective group or are not encoded correctly, the call fails.
+        Otherwise, return one if
+        log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0
+        (in F_q) and zero else.
+```
+
+Note that `k` is determined from the length of the input. Following the section on the encoding below,
+`k` is the length of the input divided by `192`. If the input length is not a multiple of `192`,
+the call fails. Empty input is valid and results in returning one.
+
+In order to check that an input is an element of `G_1`, verifying the encoding of the coordinates and checking that they satisfy the curve equation (or is the encoding of infinity) is sufficient. For `G_2`, in addition to that, the order of the element has to be checked to be equal to the group order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617`.
+
+### Definition of the groups
+
+The groups `G_1` and `G_2` are cyclic groups of prime order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617` on the elliptic curve `alt_bn128` defined by the curve equation
+`Y^2 = X^3 + 3`.
+
+The group `G_1` is a cyclic group on the above curve over the field `F_p` with `p = 21888242871839275222246405745257275088696311157297823662689037894645226208583` with generator `P1 = (1, 2)`.
+
+The group `G_2` is a cyclic group on the same elliptic curve over a different field `F_p^2 = F_p[i] / (i^2 + 1)` (p is the same as above) with generator
+```
+P2 = (
+  11559732032986387107991004021392285783925812861821192530917403151452391805634 * i +
+  10857046999023057135944570762232829481370756359578518086990519993285655852781,
+  4082367875863433681332203403145435568316851327593401208105741076214120093531 * i +
+  8495653923123431417604973247489272438418190587263600148770280649306958101930
+)
+```
```

Comment:
Mentioning this generator constant is superfluous. It is not needed to implement the precompiles. Here is a branch of the pyethereum elliptic pairing dependency that does not use this constant at all, but passes every test case: https://github.com/ethereum/py_ecc/pull/3

Reply:
What are the alternative definitions of `G_1` and `G_2`?

Reply:
Here's an attempt at `G_2` using only the constants in the py_ecc branch:

The group with order `curve_order` on the curve `y**2 * z - x**3 == b2 * z**3` where `b2 = FQ2([3, 0]) / FQ2([9, 1])` and `curve_order = 36*(BN_PARAM_U**4) + 36*(BN_PARAM_U**3) + 18*(BN_PARAM_U**2) + 6*BN_PARAM_U + 1` and `BN_PARAM_U = 4965661367192848881`.

Reply:
@cdetrio we would need to add a proof that there is only one group of that order.

Reply:
Isn't that already a [TODO](https://github.com/ethereum/EIPs/pull/212#pullrequestreview-23714624) from @chriseth?

> TODO: Check that G_2 is the only subgroup of the elliptic curve with that order.

Reply:
@cdetrio the current specification with the generator specifies a unique existence without this TODO.  Your approach requires solving this TODO.

Reply:
@cdetrio In addition, with your approach, you need to prove that a subgroup of that order exists.